{% extends "base.html" %}

{% block style %}
/*.tooltip-inner {
  background-color: #0088cc;
}*/

#image_div {
	padding-left : 5px;
}

#ideograms text,  g:not([id$='-axis'])[id^='plot'] text, g:not([id$='-axis'])[id^='plot'] > path, g:not([id$='-axis'])[id^='plot'] polygon, g:not([id$='-axis'])[id^='plot'] rectangle, g:not([id$='-axis'])[id^='plot'] circle, 
#track_0 path {
	cursor: crossor;
}

#options{
	
	padding-left : 0px;

}
.panzoom {
     -webkit-backface-visibility: initial !important;
     -webkit-transform-origin: 50% 50%;
 }
 .panel-body-gi { 
 	font-size :0.9em;
}
{% endblock %}

{% block help %}
<center><h3>Generated Circos</h3></center>
<p>Each of your generated circos will have a <b>unique ID</b> assossiated to it. This ID is a three-part-numerical code. In the event that you get an error message while generating a Circos, take note of the ID: we can trace back the problem that cause the error. You will then be able to make the appropriate change(s). </p>
<p>You can dowload the generated circos in different format.</p>
<p><ul>
  <li>By clicking on <b><span class="glyphicon glyphicon-floppy-save"></span></b> you can download the <b>svg</b> (Scalable Vector Graphics) version of your circos as currently see on your monitor i.e. if you zoom too much, you will not have a complete image, and if you apply highlights, they will be displayed</li>
  <li>If you want to access the configuration filed used to generate your circos, you can dowload a compressed <b>zip</b> package by clicking on <b><span class="glyphicon glyphicon-duplicate"></span></b>: this package contains every file used to make the image (.txt and .conf). You can then make modification(s) and upload the package to generate a new and improve Circos. <em>This option is only available for circos generated from data files i.e. not for those generated from configuration package.</em></li>
  <li>Finally, if the image is not what you wanted and would like to dispose of it, you can always delete it by clicking on <b><span class="glyphicon glyphicon-trash"></span></b>. </li>
</ul></p>

<center><h3>The Image</h3></center>
<p>If you generated the image by uploading data file(s), the Circos will have pre-defined attributes and parameters set by the interface.</p> 
<p>In avery case, the image is <b>dragable</b>: to do so, just click on the image and, while holding your click, drag it to where you want it to be. You can also find out <b>information</b> about the values represented as label(s), track(s) and links by <em>hovering</em> the Circos. For example, if you have a histogram track you can know the value of a specific bin by placing your cursor over it.</p>

<center><h3>Legend</h3></center>
The <b>legend</b> is there as a track identifier. The first thing to be displayed will be the name of your links, if you uploaded data to be represented as links. Then, all tracks will be listed from outermost to innermost, with the name you wrote while uploading your file(s).</p>
<p>To hide the legend block, just click the <em>x</em> button; to show it, click the button labelled <em>Legend</em>.</p>

<center><h3>Details</h3></center>
<p>The <b>details</b> block is there as reminder and an information holder. It contains information about the task:</p>
<p><ul>
  <li><font color="#6495ED">Type</font>: Data file(s) <em>or</em> Configuration files <em>or</em> Tabular</li>
  <li><font color="#6495ED">Description</font>: <em>personal notes</em></li>
  <li><font color="#6495ED">Date</font>: YYYY-MM-DD <em>when the circos was generated</em></li>
  <li><font color="#6495ED">ID</font>: <em>unique ID</em></li>
  <li><font color="#6495ED">ZIP</font>: <em>uploaded_package_name</em> (<b>Configuration files only</b>)</li>
  <li><font color="#6495ED">Karyotype</font>: hs <em>or</em> mm (<b>Data file(s) only</b>)</li>
  <li><font color="#6495ED">Density</font>: yes <em>or</em> no <em>is there a Gene Density track?</em> (<b>Data file(s) only</b>)</li>
  <li><font color="#6495ED"><em>links_name</em></font>: <em>link_file_name</em> (<b>Data file(s) only</b>)</li>
  <li><font color="#6495ED"><em>track_name</em></font>: <em>track_file_name</em> (<b>Data file(s) only</b>)</li>
</ul></p>
<p>To hide the details block, just click the <em>x</em> button; to show it, click the button labelled <em>Details</em>.</p>

<center><h3>Options</h3></center>
<h4><font color="#6495ED">Zoom</font></h4>
<p>You can <b>zoom</b> up to 250% of the image by adjusting the zoom range-bar. By default, the image is set to 100%. The svg won't be modified, but you can download a specific zoom as mentionned above.</p>
<h4><font color="#6495ED">Highlights</font></h4>
<p>You also have the option to <b>highligth</b> any value of your Circos. Turn the <em>Highlights</em> button <b>ON</b> and select the desired value: histogram bins, circles and tiles will be filled in red, and labels will have a red font.</p>
<p>To come back to your original Circos, click the <b>Reset</b> button: the modifications are only superficial, meaning that there are not saved as a new image. You can however download the modified image: refer to <em>Generated Circos</em></p>
<h4><font color="#6495ED">Values & Gene Information</font></h4>
<p>By turning this switch <b>ON</b>, you are able to view the values of each track directly on the circos. To do so, hover you mouse on the image and blue bubbles will appear with the <b>value</b>,  <b>chromosome</b> and <b>position</b> for track, and <b>name</b>,  <b>chromosome</b> and <b>position</b> for labels. By clicking on the desired region, a request will be sent to look for possible gene with these specific positions: if there is a match, the information will be displayed in a new box above the reset button. These are indicators.</p>
<p><em>If you are generated a circos from a configuration package, you must select the organism corresponding the displayed data: homo sapien or mouse. This option will be disabled if you are generating from data files.</em></p>
{% endblock %}

{% block header %}
<h2> Generated Circos <small>{{ unique }}</small></h2>
{% endblock %}

{% block content %}
<div class="row">
	<div class="col-md-8">
		{% if user != 'Guest' %}
		<button class="btn btn-default btn-sm pull-right" type="button" data-toggle="modal" data-target="#delete_confirm" title="Delete"><span class="glyphicon glyphicon-trash"></span></button>
		{% else %}
		<button class="btn btn-default btn-sm pull-right disabled" type="button" data-toggle="modal" data-target="#delete_confirm" title="Delete"><span class="glyphicon glyphicon-trash"></span></button>
		{% endif%}
		{% if 'ata' in type %}
			<a class="btn btn-default btn-sm pull-right" href="/download/{{ unique }}/circos.zip" title="Download ZIP"><span class="glyphicon glyphicon-duplicate"></span></a>
		{% else %}
			<a class="btn btn-default btn-sm pull-right disabled" href="/download/{{ unique }}/circos.zip" title="Download ZIP"><span class="glyphicon glyphicon-duplicate"></span></a>
		{% endif %}
		<button class="btn btn-default btn-sm pull-right" id="save_as_svg" value="" title="Download SVG"><span class="glyphicon glyphicon-floppy-save"></span></button>
	</div>
	<div class="col-md-4">
		<!-- Zoom -->
		<div class="row">
			<div class="col-md-10">
				<label for="zoom-range"><font color="#0088cc">Zoom <span id="perc" >100%</span></font></label> 
				<!-- <input type="range" id="zoom-range" min="0.5" max="2.5" value="1" step="0.1"></input> -->
		 	    <input type="range" id="zoom-range" min="50" max="450" value="100" step="1" onchange="zoom(this.value)" oninput="zoom(this.value)"></input>
			</div>		
		</div>
	</div>
</div>
<div class="row">
	<div class="col-md-8" id="image_div">
		<div class='svg' id='image_svg' display="block" margin="auto"></div>
	</div>

	<div class="col-md-4" >
		<div id="options">
		{% if 'Tabular' not in type %}
		<!-- Enable Highlights -->
		<div class="row" style="padding-top:10px;">
			<div class="col-md-5">
				<label for="highlights"><font color="#0088cc">Highlights</font></label> 
			</div>
			<div class="col-md-5">				
				<input class="disabled" id="highlights" type="checkbox" name="switch" data-handle-width="50px" data-size="mini" onChange="h()">
			</div>
		</div>
		{% endif %}

		{% if 'Tabular' not in type %}
		<div style="display:block">
		{% else %}
		<div style="display:none">
		{% endif %}
			<!-- Tooltip info -->
			<div class="row" >
				<div class="col-md-5">
					<label for="gene_info"><font color="#0088cc">Values & Gene Information</font></label> 
				</div>
				<div class="col-md-5">				
					<input id="val" type="checkbox" name="switch" data-handle-width="50px" data-size="mini" onChange="v()" >
				</div>
			</div>
		
			<div class="row">
				<div class="col-md-5">
					<label for="species"><font color="#0088cc">Gene information for: &nbsp&nbsp&nbsp</font></label> 
				</div>
				<div class="col-md-5">
					{% if 'onfig' in type %}
						<input type='radio' name='species' id='species' value='human' checked>&nbspHomo sapiens<br> 				
						<input type='radio' name='species' id='species' value='mouse'>&nbspMouse
					{% else %}
						<input type='radio' name='species' id='species' value='human' disabled>&nbspHomo sapiens<br> 				
						<input type='radio' name='species' id='species' value='mouse' disabled>&nbspMouse
					{% endif %}
				</div>
			</div>
		</div>
		</div>
		
		<!-- Hidden -->
		<div class="row" id="gene_info" style="display:none">
			<br>
			<div class="col-md-10">
				<div class="panel panel-default">
    					<div class="panel-body" >
						<button type="button" class="close" onclick="close_panel('gene_info')">&times;</button>
						<center><font color="#0088cc"><b><u> About Gene(s) </u></b></font></center>
						<div class="panel-body-gi">
						</div>
					</div>
				</div>
			</div>
		</div>
		<div class="row">
			<div class="col-md-5">
				<button id="reset" type="button" class="btn btn-default pull-left" onclick="reset()"><font color="#0088cc">Reset</font></button>
			</div>
			<div class="col-md-10">
				<hr>
			</div>
		</div>
		<br>

		<!-- hidden -->
		<div class="row" id ="show_l" style="display:none">
			<div class="row">
				<div class="col-md-10">
					<center><button class="astext btn btn-default btn-sm" onclick="show_panel('legend')">Show Legend</center>
				</div>
			</div>			
		<br>
		</div>

		<div class="row" id ="show_d" style="display:none">
			<div class="row">
				<div class="col-md-10">
					<center><button class="astext btn btn-default btn-sm" onclick="show_panel('details')">Show Details</center>
				</div>
			</div>
		</div>

		<div class="row" id="legend" style="display:block">
			<div class="col-md-10">
				<br>
				{% if lines_legend|length > 0 %}
				<div class="panel panel-default" id="myLegend">
    					<div class="panel-body">
						<button type="button" class="close" onclick="close_panel('legend')">&times;</button>
						<center><font color="#0088cc"><b><u> Legend </u></b></font></center>
						{% for l in lines_legend %}
							{{ l }}
							<br>
						{% endfor %}
		 				<br>
					</div>
				</div>
				{% endif %}
			</div>
		</div>
		{% if lines_info|length > 0 %}
		<div class="row" id="details" style="display:block">
			<div class="col-md-10">
				<br>
				<div class="panel panel-default" id="myInfo">
    					<div class="panel-body">
						<button type="button" class="close" onclick="close_panel('details')">&times;</button>
						<center><font color="#0088cc"><b><u> Details </u></b></font></center>
						{% for l in lines_info %}
							{{ l }}
							<br>
						{% endfor %}
		 				<br>
					</div>
				</div>
			</div>
		</div>
		{% endif %}
	</div>
</div>

<!-- Hidden <FORM> to submit the SVG data to the server, which will convert it to SVG/PDF/PNG downloadable file.
     The form is populated and submitted by the JavaScript below. -->
<form id="svgform" method="post" action="/download_highlights/{{ unique }}">
	<input type="hidden" id="output_format" name="output_format" value=""></input>
	<input type="hidden" id="data" name="data" value=""></input>
</form>


<!-- delete modal -->
<div class="modal fade" id="delete_confirm" role="dialog">
	<div class="modal-dialog"> 
		<div class="modal-content">
			<div class="modal-body">
				<h3><center>Deleting Circos</center></h3>
				<p>Are you sure you want to delete this Circos: </p>
				<p><pre>{% for l in lines_info %}{{ l }}{% endfor %}</pre></p><br>
				<p><em>Once deleted, a Circos and its related files can not be retrieved.</em></p>
			</div>
			<div class="modal-footer">
				<button type="button" class="btn btn-default close pull-left" data-dismiss="modal">Cancel</button>
				<a href="/delete/{{ unique }}" class="btn btn-danger pull-right" role='button'>Delete</a>
			</div>
		</div>
	</div>
</div>
{% endblock%}


{% block script %}
<script src="{{url_for('static', filename='js/jquery.panzoom.js')}}" type='text/javascript'></script>
<script src="{{url_for('static', filename='js/tooltip.js')}}" type='text/javascript'></script>
<script src="{{url_for('static', filename='js/bootstrap-switch.js')}}" type='text/javascript'></script>
<script type='text/javascript'> 

/*******************************************
****					****
****	    	INFO - LEGEND		****
****					****
********************************************/
//close panel -> gene info, info, legend
function close_panel(id) {
	console.debug('close: ' + id)
	document.getElementById(id).style.display = "none";
	if (id == 'legend') {
		document.getElementById('show_l').style.display = "block";
	}
	if (id == 'details') {
		document.getElementById('show_d').style.display = "block";	
	}
}

function show_panel(id) {
	console.debug('show: ' + id)
	document.getElementById(id).style.display = "block";
	if (id == 'legend') {
		document.getElementById('show_l').style.display = "none";		
	}
	if (id == 'details') {
		document.getElementById('show_d').style.display = "none";
	}
}

/*******************************************
****					****
****	    	Download		****
****					****
********************************************/
$("#save_as_svg").click(function() { 
	submit_download_form('svg');
});

function submit_download_form(output_format){
	//get the svg element
	var tmp = document.getElementById("image_svg");
	var svg = tmp.getElementsByTagName("svg")[0];

	// Extract the data as SVG text string
	var svg_xml = (new XMLSerializer).serializeToString(svg);

	// Submit the <FORM> to the server.
	// The result will be an attachment file to download.
	var form = document.getElementById("svgform");
	form['output_format'].value = output_format;
	form['data'].value = svg_xml ;
	form.submit();
}

/*******************************************
****					****
****	    	GET SVG 		****
****					****
********************************************/
var timesRun = 0;

getImage();

var refresh = setInterval (function(){ getImage(); }, 10000);

//callback function -> get value from asynchronous
function ajax(c, s, e, species, callback){
	var c = c.replace("chr","")
	$.ajax({
		url : "/gene_info/" + c + "/" + s + "/" + e + "/" + species,
		cache : false,
		datatype: "text",
		success: function(i){
			callback(i);
		},
	});
}

var species = document.getElementById("species").value;
var server = "http://rest.ensembl.org";

function getImage () {
	//console.debug('getImage');
	$.ajax({ 
 	   url : {% block url%}"{{url_for('get_svg', unique=unique, type=type)}}"{% endblock %},   
 	   cache : false,
 	   datatype : 'xml',

 	   success : function(content) {
 	   		//console.debug(++timesRun);		
 	   		if (content=='' | content ==undefined) {
 	   			{% block display %}	
				$('#image_svg').html('<div style="padding:10px;"><center><i class="fa fa-spinner fa-spin fa-3x"></i></center></div>');
				{% endblock%} 	   		
			}

			else if (content == 'error') {
				$("#image_svg").html('<div class="alert alert-danger" role="alert"><span class="glyphicon glyphicon-exclamation-sign" aria-hidden="true"></span><em>An error has occured while generating your circos. Please check your files and try again</em></div>');
				$("#info").html('');
				clearInterval(refresh);
			}

 	   		else {  
				$("#image_svg").css('border', 'solid #eee 1px');
				$("#image_svg").html(content);
                var svg = $('#image_svg').find('svg')[0];
				svg.setAttribute('width', "100%");
				svg.setAttribute('height', "100%");
				svg.setAttribute('viewBox', "-300 -300 3500 3500");
				clearInterval(refresh);						
			}
        },
       error : function(e) { console.debug(e);},
      });
}

/*******************************************
****					****
****	    	INTERACTIVITY 		****
****					****
********************************************/
$(document).ready(function() {
	$("[name='switch']").bootstrapSwitch();	
});


//$('.tooltip').tooltip('hide');

var panzoom =  $("#image_svg").panzoom({transition:false});//,
/*						$zoomRange: $('#zoom-range'),
						increment : 0.1,						
						minsScale : 0.5,
						maxScale : 5,
						rangeStep :0.1});

$(panzoom).on('panzoomzoom', function(e, panzoom, scale, opt) {
    $("#perc").html((scale*100).toFixed(0) +"%");

});
$(panzoom).on('panzoomend', function(e, panzoom, m, changed) {
	if (changed) {

	}
};
*/
// zoom with wheel mouse  // jquery.panzoom.js example


// Not using the zoom of panzoom 
$(panzoom).parent().on('mousewheel', function( e ) {
            e.preventDefault();
            val = parseInt($("#zoom-range").val());
            increment = parseInt($("#zoom-range").attr('step'))*5;
            var delta = e.delta || e.originalEvent.wheelDelta;
            var zoomOut = delta ? delta < 0 : e.originalEvent.deltaY > 0;
            var zoomValue = zoomOut ? val - increment : val + increment
            zoom(zoomValue);
            $("#zoom-range").val(zoomValue);
            $("#perc").text(zoomValue+'%');
});

// Not using the zoom of panzoom 
 function zoom(val){
 	//$('.tooltip').tooltip('hide');
 	var svg = $('#image_svg').find('svg')[0];
 	var v = val + "%"
 	var vb = [val * -3, val * -2.5, 3000 + (5 * val), 3000 + (5 * val)]
 	var n =  val < 100 ? -1 : 1 ; 
 	$("#perc").text(v);
 	svg.setAttribute('width', v);
 	svg.setAttribute('height', v);
 	svg.setAttribute('viewBox', vb.join(" "));
 	
 	// Need to find the right formula... Would like to keep the image centered after zoom
 	//var m  = 'matrix(1,0,0,1,'+ n*vb[0]+','+vb[1]+')'
	//$("#image_svg").css('webkit-transform', m);
 	//$("#image_svg").css('transform', m);
 	
 };

function reset() {
	document.location.reload(true);
};

//highlight -> ON/OFF
var emphasized = {}

var emphasize_me = function (el, val, opts) {
	if (emphasized[val]==undefined) {
		var style = el.getAttribute("style");
		emphasized[val] = style;	
		$(el).css(opts);
	}
};

var emphasize_ml = function(el, val) { 
	$(el).attr("style", emphasized[val]);
	delete emphasized[val];
};

var emphasize_click = function(el, val) {
	if (el.getAttribute("id") == "e") {		// back un-emphasized
		$(el).removeAttr('id');
		$(el).on('mouseleave.h', function() {emphasize_ml(el, val)});
	} else {
		$(el).attr('id', 'e');	
		$(el).off('mouseleave.h');		// keep emphasized
	}			
};

function h(){

	if(document.getElementById('highlights').checked){
		$("#ideograms text, g:not([id$='-axis'])[id^='plot'] text").on('mouseenter.h',  function() {
				var opts = {'fill':'crimson'};
				var x = this.getAttribute("x");
				emphasize_me(this, x, opts);
		});
		$("#ideograms text, g:not([id$='-axis'])[id^='plot'] text").on('mouseleave.h',  function() {
				var x = this.getAttribute("x");
				emphasize_ml(this, x);
		});
		$("#ideograms text, g:not([id$='-axis'])[id^='plot'] text").on('click.h',  function() {
				var x = this.getAttribute("x");
				emphasize_click(this, x);
		});

		// links only!! //and plot-axis and g[id^='plot'][id$='-axis'] > path
		$("#track_0 path").on('mouseenter.h', function() {
				opts = {'stroke':'crimson', 'stroke-width':'2px'};
				if (this.parentNode.id == "track_0") {
					opts = {'stroke' : 'crimson', 'stroke-width' : '7.0'};
				}
				var d = this.getAttribute("d");	
				emphasize_me(this, d, opts);
		});
		$("#track_0 path").on('mouseleave.h', function() {
				var d = this.getAttribute("d");
				emphasize_ml(this, d);
		});
		$("#track_0 path").on('click.h', function() {
				var d = this.getAttribute("d");
				emphasize_click(this, d);
		});
			
		//plots
		$("g:not([id$='-axis'])[id^='plot'] > path").on('mouseenter.h', function(){
				var d = this.getAttribute("d");
				opts = {'fill' : 'crimson'};
				if (this.getAttribute("class") == "tile" || this.getAttribute("class")=='histogram') {
					opts = {'fill' : 'crimson', 'stroke' : 'crimson', 'stroke-width' : '3.0'};
				}
				emphasize_me(this, d, opts);
		});
		$("g:not([id$='-axis'])[id^='plot'] > path").on('mouseleave.h', function(){
				var d = this.getAttribute("d");
				emphasize_ml(this, d);
		});
		$("g:not([id$='-axis'])[id^='plot'] > path").on('click.h', function(){
				var d = this.getAttribute("d");
				emphasize_click(this, d);
		});


		//scatter plots -> polygons
		$("g:not([id$='-axis'])[id^='plot'] polygon, g:not([id$='-axis'])[id^='plot'] rectangle, g:not([id$='-axis'])[id^='plot'] circle").on('mouseenter.h', function() {
				var cx = this.getAttribute("cx");
				opts = {'fill':'crimson', 'stroke':'fireBrick','stroke-width':'3.0'};
				emphasize_me(this, cx, opts);
		});
		$("g:not([id$='-axis'])[id^='plot'] polygon, g:not([id$='-axis'])[id^='plot'] rectangle, g:not([id$='-axis'])[id^='plot'] circle").on('mouseleave.h', function() {
				var cx = this.getAttribute("cx");
				emphasize_ml(this, cx);
		});
		$("g:not([id$='-axis'])[id^='plot'] polygon, g:not([id$='-axis'])[id^='plot'] rectangle, g:not([id$='-axis'])[id^='plot'] circle").on('click.h', function() {
				var cx = this.getAttribute("cx");
				emphasize_click(this, cx);
		});		



	} 
	else {
		$("#ideograms text,  g:not([id$='-axis'])[id^='plot'] text").off("mouseenter.h mouseleave.h click.h");
		$("g:not([id$='-axis'])[id^='plot'] > path, g:not([id$='-axis'])[id^='plot'] polygon").off("mouseenter.h mouseleave.h click.h");
		$("g:not([id$='-axis'])[id^='plot'] rectangle, g:not([id$='-axis'])[id^='plot'] circle").off("mouseenter.h mouseleave.h click.h");
		$("#track_0 path").off("mouseenter.h mouseleave.h  click.h");
	}
}



var build_gene_info = function (o, op) {
	
	var n = o.external_name;
	var i = o.gene_id;
	var f = o.strand;
	var b = o.biotype.replace("_", " ");
	var d = o.description==null ? o.description : o.description.split('[')[0];
	
	var ebLink = "http://useast.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g="+i;
  	var gcLink = 'http://www.genecards.org/cgi-bin/carddisp.pl?gene='+n+'&search='+n+'';
  	var egLink = 'http://www.ncbi.nlm.nih.gov/gene?cmd=search&term='+i;
  	//var wkLink = 'http://en.wikipedia.org/wiki/'+n;
  	
  	var links = {'Ensembl':ebLink, 'GeneCards':gcLink, 'EntrezGene': egLink}; //, 'Wikipedia': wkLink};
  	var infos = ['Ensembl','EntrezGene']

  	var l = '<center><ul class="breadcrumb">';
	_.each(_.pairs(links), function(p) { l = l + '<li><a href="'+p[1]+'" target="_blank"> <strong>'+p[0]+'</strong></a></li>'; }); 
	l = l + '</ul></center>';

	$(".panel-body-gi").empty().append("<div class='row'></div>");
	$(".panel-body-gi .row").append("<div class='col-md-12' style='padding-top:5px;'>"+ "<i><b>For position: </b>"+op.s+"-"+op.e+" (<b>value: </b>"+op.val+")</i></div>");
	//"<br><b>ID: </b>"+ i +
	$(".panel-body-gi .row").append("<div class='col-md-6' style='padding-top:5px;'><b>Gene: </b>" + n +  "<br><b>Biotype: </b>" + b + "</div>");

	s = "<div class='col-md-6' style='padding-top:5px;'>"+ "<b>Chr: </b>" + op.chr + "<br><b>Strand: </b>" + f;
	if (o.Location!= undefined){
		s = s + "<br><b>Location: </b>" + o.Location;
	}
	$(".panel-body-gi .row").append(s+"</div>");
	
	$(".panel-body-gi .row").append("<div class='col-md-12'>"+"<b>Description: </b>" + d+"<br></div>");
	if (o['Other Aliases']!= undefined) {
		$(".panel-body-gi .row").append("<div class='col-md-12'>"+"<b>Aliases: </b>" + o['Other Aliases']+"<br></div>"); }
	$(".panel-body-gi .row").append("<div class='col-md-12' style='padding-top:10px;'>"+l+"</div>");

} 

var gene_data = {};
var o = null;
var add_gene_information = function (op) {
	$(".panel-body-gi").empty().append('<center><div style="padding-top:10px;"><i class="fa fa-spinner fa-spin fa-2x"></i></div></center>');
    var chr = op.chr;
    var s = op.s;
    var e = op.e; 
    var val = op.val;
    var url1 = server+"/map/"+species+"/GRCh37/"+String(chr.replace("chr",""))+":"+String(s)+".."+String(e)+":1/GRCh38?;content-type=application/json";
   	
  	var urlNCBI  = 'http://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=gene&id=_id_&retmode=txt'; 
  	var urlEnsemblId = 'http://rest.ensembl.org/xrefs/id/_id_?content-type=application/json';
    
    if (gene_data[val]){
    	build_gene_info (gene_data[val], op);
    	return
    }

	$.ajax({url: url1}).then(function(data1) {
		
		// No mapping found between genome version
		if (data1.mappings.length == 0){	
			$(".panel-body-gi").empty().append("<div class='row'></div>");
			$(".panel-body-gi .row").append("<div class='col-md-12' style='padding-top:5px;'>"+ "<i><b>For position: </b>"+s+"-"+e+" (<b>value: </b>"+val+")</i></div>");
			$(".panel-body-gi .row").append("<div class='col-md-6'><em>This position does not correspond to a gene</em></div><div class='col-md-6'><b>Chr: </b>" + chr +"</div>");
			return;
		}
		// multiple mapping found
		if (data1.mappings.length > 1) {
			$(".panel-body-gi").empty().append("<div class='row'></div>");
			$(".panel-body-gi .row").append("<div class='col-md-12' style='padding-top:5px;'>"+ "<i><b>For position: </b>"+s+"-"+e+" (<b>value: </b>"+val+")</i></div>");
			$(".panel-body-gi .row").append("<div class='col-md-6'><em>There are several genes within the selected position</em></div><div class='col-md-6'><b>Chr: </b>" + chr + "</div>");
			return;
		}

		
		// Only one mapping found
		else {
			var ms = data1["mappings"][0]["mapped"]["start"];
    		var me = data1["mappings"][0]["mapped"]["end"];
			var url2 = server+"/overlap/region/"+species+"/"+String(chr.replace("chr",""))+":"+String(ms)+"-"+String(me)+"?feature=gene;content-type=application/json";
			
			$.ajax({ url: url2 }).then(function(data2) {
			// Gene not found
			if (data2.length == 0){
				$(".panel-body-gi").empty().append("<div class='row'></div>");
				$(".panel-body-gi .row").append("<div class='col-md-12' style='padding-top:5px;'>"+ "<i><b>For position: </b>"+s+"-"+e+" (<b>value: </b>"+val+")</i></div>");
				$(".panel-body-gi .row").append("<div class='col-md-6'><em>This position does not correspond to a gene</em></div><div class='col-md-6'><b>Chr: </b>" + chr  + "</div>")
			}
			else {
				
				//console.debug(url1);
				//console.debug(url2);
				o  = data2[0];
				if (data2.length > 1 ) {
					if (isNaN(val)) {
						for (i = 0; i < data2.length; i++) { 
						
    						if (data2[i].external_name==val) { o = data2[i];  break;}
    					}
					}
					else {
						$(".panel-body-gi").empty().append("<div class='row'></div>");
						$(".panel-body-gi .row").append("<div class='col-md-12' style='padding-top:5px;'>"+ "<i><b>For position: </b>"+s+"-"+e+" (<b>value: </b>"+val+")</i></div>");
						$(".panel-body-gi .row").append("<div class='col-md-6'><em>There are " + data2.length + " genes within the selected position</em></div><div class='col-md-6'><b>Chr: </b>" + chr + "</div>");
						 return;
					}
				}
					
				
				$.ajax({
					url:urlEnsemblId.replace('_id_', o.gene_id), 
					success : function(r) { 
						console.debug(r);
						eg = _.where(r, {dbname:"EntrezGene"});
						if (eg.length ==1 ){
							egid = eg[0].primary_id;
              			}
              			else {
              				gene_data[val] = o;
							build_gene_info (o, op);
							return;
              			}
              			$.ajax({url : urlNCBI.replace('_id_', egid), 
              					success : function(r){ 
              			
              						r = r.split('\n');
              						_.each(r, function(a) {
           							r1 = a.split(':');
           					
           							if (!(_.isUndefined(r1[0])) & !(_.isUndefined(r1[1])) ){
           		 						if (r1[1].search('Name')!=-1) {
           									r1[1] = r1[1].replace('and Name', '');
           									o['Name']= r1[2].replace(/ /g,'');;
           								}
           								if (r1[1].search('Location')!=-1) {
           									r1[1] = r1[1].replace('; Location', "");
           									o['Location']= r1[2].replace(/ /g,'');;
           								}
           		 						o[r1[0]]= r1[1].replace(/ /g,'');;
           		 					}
           							});
           				
              		 			}, 
              		 			complete:  function() { 
              		 				gene_data[val] = o;
									build_gene_info (o, op);
								}
              			});
              		}
              	});

              	
			}
		});									
	}
 });
};

var get_data = function(t) {
	var val = t.getAttribute("data-value").replace(/'/g,"");
	var chr = (t.getAttribute("data-chr")).replace(/'/g,"");
	var s = t.getAttribute("data-start");
	var e = t.getAttribute("data-end");
	return {val:val, chr:chr, s:s, e:e}
};

var get_tooltip_title = function (op) {
	return '<b>' + op.val + '</b><br>' + op.chr + ': ' + op.s + '-' + op.e 
};

//values + info -> ON/OFF
function v(){

	var tooltip_opts = {title: "",
						container: 'body', 
						placement: "auto left", 
						trigger: 'hover', 
						html:true};

	if(document.getElementById('val').checked){
		
		$("g[id^='plot'] path, g[id^='plot'] text, g[id^='plot'] rectangle, g[id^='plot'] polygon, g[id^='plot'] circle").each( function(i,v){
			
			id = $(this).parent()[0].id;
			if (id.search('axis')==-1){
				var op = get_data (this);
				tooltip_opts.title = get_tooltip_title (op);
				tooltip_opts.viewport = this;
				$(this).tooltip(tooltip_opts);

				$(this).on("click.v", function(){ 
					$(this).on("mouseleave.v", function(){ 
						$(this).tooltip("hide") 
					});
					document.getElementById("gene_info").style.display = "block";	
					add_gene_information (op);
					
				});
			}
		});
		
		// links
		$("#track_0 path").each(function(i,v) {
			var chr1 = (this.getAttribute("data-chr1")).replace(/'/g,"");
			var s1 = this.getAttribute("data-start1");
			var e1 = this.getAttribute("data-end1");
			var chr2 = (this.getAttribute("data-chr2")).replace(/'/g,"");
			var s2 = this.getAttribute("data-start2");
			var e2 = this.getAttribute("data-end2");
			
			$(this).tooltip({title: chr1 + ': ' + s1 + '-' + e1 + '<br>to<br>' + chr2 + ': ' + s2 + '-' + e2 , container: 'body', placement: "auto left", trigger:'hover', html:'true', viewport:this});
			$(this).on("click.v", function(){
				$(this).on("mouseleave", function(){
					$(this).tooltip("hide")
				});
			});
		});
	} 
	else {
		
		$("g[id^='plot'] path, g[id^='plot'] text, g[id^='plot'] circle, g[id^='plot'] rectangle, g[id^='plot'] polygon, #track_0 path").each(function(i,v) {
			id = $(this).parent()[0].id;
			if (id.search('axis')==-1){
				$(this).tooltip('destroy');
				$(this).off("click.v");
			}
		});
	}
}


</script>
{% endblock %}
